include/flatbuffers/flatbuffers.h

```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
-  virtual uint8_t *allocate(size_t size) const { return new uint8_t[size]; }
-  virtual void deallocate(uint8_t *p) const { delete[] p; }
+  virtual ~Allocator() {}
+
+  // Allocate `size` bytes of memory.
+  virtual uint8_t *allocate(size_t size) = 0;
+
+  // Deallocate `size` bytes of memory at `p` allocated by this allocator.
+  virtual void deallocate(uint8_t *p, size_t size) = 0;
+
+  // Reallocate `new_size` bytes of memory, replacing the old region of size
+  // `old_size` at `p`. In contrast to a normal realloc, this grows downwards,
+  // and is intended specifcally for `vector_downward` use.
+  virtual uint8_t *reallocate_downward(
+      uint8_t *old_p, size_t old_size, size_t new_size) = 0;
+
+  // Release myself to the caller by creating a new
+  // `std::unique_ptr<Allocator>` move of myself and resetting my own state.
+  virtual std::unique_ptr<Allocator> release() = 0;
 };
 
+class DefaultAllocator : public Allocator {
+ public:
+  virtual uint8_t *allocate(size_t size) FLATBUFFERS_OVERRIDE {
+    return new uint8_t[size];
+  }
+
+  virtual void deallocate(uint8_t *p, size_t /* unused */) FLATBUFFERS_OVERRIDE {
+    delete[] p;
+  }
+
+  virtual uint8_t *reallocate_downward(
+      uint8_t *old_p, size_t old_size, size_t new_size) FLATBUFFERS_OVERRIDE {
+    assert(new_size > old_size);  // vector_downward only grows
+    uint8_t* new_p = new uint8_t[new_size];
+    memcpy(new_p + (new_size - old_size), old_p, old_size);
+    delete[] old_p;
+    return new_p;
+  }
+
+  virtual std::unique_ptr<Allocator> release() FLATBUFFERS_OVERRIDE {
+    // no state to transfer, so just release a new instance
+    return std::unique_ptr<Allocator>(new DefaultAllocator());
+  }
+};
+
+#ifndef FLATBUFFERS_CPP98_STL
+// BufferDeleter is specific to the `release` functionality, and allows the
+// user to safely gain ownership of a buffer and its associated buffer.
+class BufferDeleter {
+ public:
+  BufferDeleter(std::unique_ptr<Allocator> allocator, uint8_t *buf, size_t size)
+      : allocator_(std::forward<std::unique_ptr<Allocator>>(allocator)),
+        buf_(buf),
+        size_(size) {
+    assert(allocator_);
+  }
+
+  BufferDeleter(BufferDeleter &&other)
+      : allocator_(std::forward<std::unique_ptr<Allocator>>(other.allocator_)),
+        buf_(other.buf_),
+        size_(other.size_) {
+    assert(allocator_);
+  }
+
+  inline void operator()(uint8_t * /* unused */) {
+    allocator_->deallocate(buf_, size_);
+  }
+
+  // must stay at end of public section, may change access to private
+  FLATBUFFERS_DELETE_FUNC(BufferDeleter(const BufferDeleter& other))
+  FLATBUFFERS_DELETE_FUNC(BufferDeleter& operator=(const BufferDeleter& other))
+
+ protected:
+  std::unique_ptr<Allocator> allocator_;
+  uint8_t *buf_;
+  size_t size_;
+};
+
+// Pointer to relinquished memory.
+#if (!defined(_MSC_VER) || _MSC_VER > 1600)
+typedef std::unique_ptr<uint8_t, BufferDeleter> unique_ptr_t;
+#else  // begin hack for MSVC 16
+class unique_ptr_t {
```

Comment:
Not necessarily the compiler config stuff.. but this emulation code is only for a small number of users so could be tucked in a corner.